tests/main/snap-quota-groups/task.yaml

```diff
@@ -0,0 +1,76 @@
+summary: Functional test for quota-related snap commands.
+
+details: |
+  Functional test for snap quota group commands ensuring that they are 
+  effective in practice.
+
+prepare: |
+  if os.query is-trusty; then
+    exit 0
+  fi
+  snap install go-example-webserver jq remarshal
+  snap set system experimental.quota-groups=true
+  tests.cleanup defer snap unset system experimental.quota-groups
+
+execute: |
+  if os.query is-trusty; then
+    # just check that we can't do anything with quota groups on trusty, systemd
+    # there is 204, but we need at least 205 for slice units
+
+    snap quota foobar --max-memory=1MB 2>&1 | MATCH "systemd version too old: snap quotas requires systemd 205 and newer \(currently have 204\)"
+    exit 0
+  fi
+
+  echo "Create a group with a snap in it"
+  snap quota group-one --max-memory=100MB go-example-webserver
+
+  echo "The systemd slice should be active now"
+  sliceName="snap.$(systemd-escape --path group-one).slice"
+  systemctl show --property=ActiveState "$sliceName" | MATCH "ActiveState=active"
+
+  echo "The service should also still be active"
+  snap services go-example-webserver.webserver | MATCH "go-example-webserver.webserver\s+enabled\s+active"
+
+  echo "The systemd slice should have one process in it now"
+  cgroupProcsFile="/sys/fs/cgroup/memory/$sliceName/snap.go-example-webserver.webserver.service/cgroup.procs"
+  wc -l < "$cgroupProcsFile" | MATCH 1
+  SERVER_PID=$(cat "$cgroupProcsFile")
+
+  echo "And that process is the main PID for the example webserver"
+  systemctl show --property=MainPID snap.go-example-webserver.webserver.service | MATCH "MainPID=$SERVER_PID"
+
+  echo "And the service is in the Control Group for the slice"
+  test "$(systemctl show --property=ControlGroup snap.go-example-webserver.webserver.service)" = "ControlGroup=/$sliceName/snap.go-example-webserver.webserver.service"
+
+  # TODO: check that systemctl says the memory usage is the same as "snap quota group-one"
+
+  echo "Removing the quota will stop the slice and the service will be restarted"
+  snap remove-quota group-one
+  systemctl show --property=MainPID snap.go-example-webserver.webserver.service | NOMATCH "MainPID=$SERVER_PID"
+  snap services go-example-webserver.webserver | MATCH "go-example-webserver.webserver\s+enabled\s+active"
+
+  echo "And the service is not in a slice anymore"
+  systemctl show --property=ControlGroup snap.go-example-webserver.webserver.service | NOMATCH "/$sliceName/snap.go-example-webserver.webserver.service"
+
+  echo "Creating a quota with a very small memory limit results in the service being unable to start"
+  snap quota too-small --max-memory=1MB go-example-webserver
```

Comment:
Random remark, I think we want to enforce reasonable minimum here eventually, so we might need to inflate the memory usage in the test server at some point, but it's fine for now of course.

Reply:
Yeah, when we do that, I think we will notice this test failing and can fix it up then. I had a try at writing a simple memory hog service in bash, but it either a) always grabbed too much memory and died and couldn't be controlled or b) somehow paged/cached/freed some of it's memory so it wasn't killed by systemd when I tried to make it use more memory and get killed. I think we may need to write something simple in Go or maybe Python which hard allocates memory and uses it in a way we can control to test the OOM killing properly

Reply:
That being said, in the TODO comment I left in the quota implementation about there being a technical minimum, I still think there is a technical minimum, but I think that minimum is probably smaller than 1 MiB, so we maybe can continue to get away with using this snap for this sort of test